aten/src/ATen/native/LossNLL.cpp

```diff
@@ -2,10 +2,48 @@
 #include <ATen/AccumulateType.h>
 #include <ATen/Dispatch.h>
 #include <ATen/Parallel.h>
+#include <ATen/TensorMeta.h>
 #include <ATen/TensorUtils.h>
 #include <ATen/native/cpu/utils.h>
 
 namespace at {
+namespace meta {
+TORCH_META_FUNC(nll_loss_forward)
+(const Tensor& self,
+ const Tensor& target,
+ const OptionalTensorRef weight_opt,
+ int64_t reduction,
+ int64_t ignore_index) {
+  const Tensor& weight = weight_opt.getTensorRef();
+
+  TORCH_CHECK(
+      self.dim() > 0 && self.dim() <= 2, "input tensor should be 1D or 2D");
+  TORCH_CHECK(
+      target.dim() == 1,
+      "1D target tensor expected, multi-target not supported");
+  TORCH_CHECK(
+      self.size(0) == target.size(0),
+      "size mismatch (got input: ",
+      self.sizes(),
+      ", target: ",
+      target.sizes(),
+      ")")
+
+  const auto n_classes = self.size(-1);
+
+  TORCH_CHECK(
+      !weight.defined() || (weight.dim() == 1 && weight.numel() == n_classes),
+      "weight tensor should be defined either for all ",
+      n_classes,
+      " classes or no classes"
+      " but got weight tensor of shape: ",
+      weight.sizes());
+
+  set_output(0, {0}, self.options());
```

Comment:
This looks pretty dodgy, surely the zeroth output isn't always a zero size tensor.

Reply:
Given that this kernel is computing loss, I think the output is a scalar? This is the old implementation of the non-out variant:

```
  auto output = at::empty({0}, self.options());
  auto total_weight = at::empty({0}, self.options());
  at::native::nll_loss_forward_out_cpu(
      self, target, weight, reduction, ignore_index, output, total_weight);
  return std::make_tuple(output, total_weight);
```

The `{}` size for the second output comes from the unconditional `total_weight.resize_({})` in the old implementation `nll_loss_forward_out_cpu_template`.

Reply:
If the output was a scalar, the output size would be specified as `{}`, not `{0}`. My guess is `nll_loss_forward_out_cpu` is internally doing a resize.

Reply:
You're right, I found the code that resizes `output` in `nll_loss_out_frame` (which is called by `nll_loss_forward_out_cpu_template`). All the other kernels I've ported did all their resizing in `{kernel_name}_cpu_template`, not `{kernel_name}_frame` so I didn't think to look there. I'll fix in a bit after doing some PSC